Integration tests for input detection cases in `/api/v1/task/server-streaming-classification-with-text-generation`

mdevino opened the pull request:
This addreses #299.

* The list of implemented test cases can be found on #299 and were implemented on `tests/streaming_classification_with_gen.nlp`. Besides, tests to isolated chunker and generation calls are present in `tests/chunker.rs` and `tests/generation_nlp.rs`, respectively.
* As #298 got outdated and wasn't merged until now, I closed it and favor of this PR. The tests for it are present in `tests/detection_content.rs`
* Created `tests/common` to shared reusable code among test crates. The structure is as follows:
  * `tests/common/chunker.rs` hosts code associated to chunkers.
  * `tests/common/detectors.rs` hosts code associated to detectors.
  * `tests/common/errors.rs` currently contains structs representing errors returned from detectors and orchestrator. The idea behind having these classes instead of using already existing ones under `/src` is to make API changes more evident (as tests would fail due to deserialization issues in case of API changes).
  * `tests/common/orchestrator.rs` hosts code associated to the orchestrator, and also contains a test server for it.
  * `tests/common/mod.rs` declares the submodules and hosts code that is not associated to any of the above.
* Updated orchestrator test config file:
  * Chunkers: added `sentence_chunker`
  * Detectors: added `angle_brackets` detector with two configs - one associed to whole_doc_chunker and one associated to `sentence_chunker`.
* Dependency changes:
  * Updated dependencies to a later version (latest at the time of the commit).
  * Added mocktail and test-log as dev-dependencies, and removed tracing-test.

gkumbhat reviewed (commented):
Couple of high level points:
1. Can you please add more details in the PR description. The current description doesn't seem to be matching all the changes proposed in this PR
2. Can we move all the integration tests to `tests/integration` that way they don't get mixed with unit tests

mdevino commented on Cargo.toml:
```diff
@@ -28,27 +29,53 @@ http-body = "1.0"
+hyper-rustls = { version = "0.27.5", features = ["ring"] }
 hyper-timeout = "0.5.2"
-hyper-util = { version = "0.1.10", features = ["server-auto", "server-graceful", "tokio"] }
+hyper-util = { version = "0.1.10", features = [
```
Linter.

mdevino commented on tests/canary_test.rs:
```diff
@@ -25,17 +48,12 @@ async fn shared_state() -> Arc<ServerState> {
 /// superficially testing the client health endpoints on the orchestrator is accessible
 /// and when the orchestrator is running (healthy) all the health endpoints return 200 OK.
 /// This will happen even if the client services or their health endpoints are not found.
-#[traced_test]
```
The crate `tracing_test` crate is intended to be used mostly to assert something is logged, while `test_log` is more suited for actual logging (and also respects the `RUST_LOG` environment variable).

declark1 commented:
> > Can we move all the integration tests to tests/integration that way they don't get mixed with unit tests
> 
> @gkumbhat I could, but don't we implement unit tests in the appropriate files in /src?

FYI - in Rust `/tests` is for integration tests only (as it's external to `/src`), while unit tests are placed in modules under `/src`

mdevino commented on tests/chunker.rs:
```diff
@@ -0,0 +1,99 @@
+        MockPath::new(Method::POST, CHUNKER_UNARY_ENDPOINT),
+        Mock::new(
+            MockRequest::pb(ChunkerTokenizationTaskRequest {
+                text: "Hi there! how are you? I am great!".into(),
```
Done.

declark1 commented:
> tests/common/errors.rs currently contains structs representing errors returned from detectors and orchestrator. The idea behind having these classes instead of using already existing ones under /src is to make API changes more evident (as tests would fail due to deserialization issues in case of API changes).

I suggest that we just use `anyhow::Error` for tests, which all errors can be converted into similar to `Box<dyn std::error::Error>`

mdevino commented:
> > tests/common/errors.rs currently contains structs representing errors returned from detectors and orchestrator. The idea behind having these classes instead of using already existing ones under /src is to make API changes more evident (as tests would fail due to deserialization issues in case of API changes).
> 
> I suggest that we just use `anyhow::Error` for tests, which all errors can be converted into similar to `Box<dyn std::error::Error>`

@declark1 Would this enable accessing fields inside the error? I ask because I use `OrchestratorError.code` and `.details` for assertions.

declark1 commented:
> > > tests/common/errors.rs currently contains structs representing errors returned from detectors and orchestrator. The idea behind having these classes instead of using already existing ones under /src is to make API changes more evident (as tests would fail due to deserialization issues in case of API changes).
> > 
> > 
> > I suggest that we just use `anyhow::Error` for tests, which all errors can be converted into similar to `Box<dyn std::error::Error>`
> 
> @declark1 Would this enable accessing fields inside the error? I ask because I use `OrchestratorError.code` and `.details` for assertions.

Yea true, I forgot that you need to assert the specific error code. Can't you just use the existing `server::Error` and `client::Error` types?

gkumbhat commented on tests/streaming_classification_with_gen_nlp.rs:
```diff
@@ -0,0 +1,1185 @@
+        .send()
+        .await?;
+
+    // // Example showing how to create an event stream from a bytes stream.
```
nit: Is this example here to show how to write even stream involving test? In that case, can we move this at the top and make it more clearer?

Since each test is lengthy, it would be good for it not take up extra lines with such examples. Additionally, it would be not clear why this code is commented out

mdevino replied:
I've used this piece of code to troubleshoot a few test cases. I've moved it to the top of the file with a suggestion on how to use it.

gkumbhat commented on tests/streaming_classification_with_gen_nlp.rs:
```diff
@@ -0,0 +1,1185 @@
+pub mod common;
+
+#[test(tokio::test)]
+async fn test_no_detectors() -> Result<(), anyhow::Error> {
```
can you add description of what this test is doing?

mdevino replied:
Done. I've added a short description before every test (except the ones in `tests/detection_content.rs`, which will be addressed in #314).

gkumbhat commented on tests/streaming_classification_with_gen_nlp.rs:
```diff
@@ -0,0 +1,1185 @@
+    assert!(messages.len() == 3);
+    assert!(messages[0].generated_text == Some("I".into()));
+    assert!(messages[1].generated_text == Some(" am".into()));
+    assert!(messages[2].generated_text == Some(" great!".into()));
```
can we also assert / check that the detection block is not available, showing "no detections"

mdevino replied:
Done.

gkumbhat commented on tests/streaming_classification_with_gen_nlp.rs:
```diff
@@ -0,0 +1,1185 @@
+    assert!(messages[0].generated_text == Some("I".into()));
+    assert!(messages[1].generated_text == Some(" am".into()));
+    assert!(messages[2].generated_text == Some(" great!".into()));
+
```
same here, regarding checking "detection" block

mdevino replied:
Done.

gkumbhat commented on tests/streaming_classification_with_gen_nlp.rs:
```diff
@@ -0,0 +1,1185 @@
+
+#[test(tokio::test)]
+async fn test_input_detector_returns_404() -> Result<(), anyhow::Error> {
+    let detector_name = DETECTOR_NAME_ANGLE_BRACKETS_WHOLE_DOC;
```
same here about the description

mdevino replied:
Done.

gkumbhat commented:
> I could, but don't we implement unit tests in the appropriate files in /src?
@mdevino you are right. I changed the repos and missed that the unit tests are embedded in this one.

mdevino commented on tests/chunker.rs:
```diff
@@ -0,0 +1,100 @@
+async fn test_isolated_chunker_unary_call() -> Result<(), anyhow::Error> {
+    // Add detector mock
+    let chunker_id = "sentence_chunker";
+    let input_test = "Hi there! how are you? I am great!";
```
Done.

mdevino commented on tests/common/orchestrator.rs:
```diff
@@ -0,0 +1,237 @@
+        Self::initialize_orchestrator_server(port, health_port, config).await
+    }
+
+    async fn initialize_generation_server(
```
Sure. I'll raise a new issue for it.

mdevino commented on tests/test.config.yaml:
Done.

mdevino commented on tests/generation_nlp.rs:
```diff
@@ -0,0 +1,102 @@
+        .await?;
+
+    // Collect stream results as array for assertion
+    let result = response.map(Result::unwrap).collect::<Vec<_>>().await;
```
I don't see the benefit on this change. Could you elaborate?

mdevino commented on tests/detection_content.rs:
Could you point at what you're calling boilerplate?

Most lines in a text body are related to setting mock request/responses. I don't see this going away with running a single orchestrator instance.

In addition, the orchestrator configuration would need to account for all tests. This adds the complexity of having to know all existing mocks to add a new test, otherwise it could impact existing ones.

I strongly advise against reusing the orchestrator server.

gkumbhat commented on tests/test_config.yaml:
```diff
@@ -0,0 +1,35 @@
+      hostname: localhost
+    type: sentence
+detectors:
+  test_detector:
```
where is this one getting used ?

mdevino replied:
This detector was already there when I started writing tests. I've removed that and tests keep passing, so I'll push the change to remove it.

gkumbhat commented on tests/streaming_classification_with_gen_nlp.rs:
```diff
@@ -0,0 +1,1212 @@
+                contents: vec!["This should return a 500".into()],
+                detector_params: DetectorParams::new(),
+            }),
+            MockResponse::json(&expected_detector_error).with_code(StatusCode::NOT_FOUND),
```
the error code here doesn't match the error code in the body, i.e `404` (not found) vs `500`. This is pointing out that we give precedence to body than the header of the request, which is fine, but generally these 2 codes should be same, so this test is looking odd

mdevino replied:
Fixed.

mdevino commented on tests/generation_nlp.rs:
```diff
@@ -0,0 +1,101 @@
+
+/// Asserts that the NlpClient correctly invokes the streaming endpoint.
+#[test(tokio::test)]
+async fn test_nlp_streaming_call() -> Result<(), anyhow::Error> {
```
I can remove the client tests. Their main intent was to make sure I was mocking the calls correctly.

mdevino commented on tests/common/orchestrator.rs:
```diff
@@ -0,0 +1,235 @@
+use bytes::Bytes;
+use eventsource_stream::{EventStream, Eventsource};
+use fms_guardrails_orchestr8::{config::OrchestratorConfig, orchestrator::Orchestrator};
+use futures::{stream::BoxStream, Stream, StreamExt};
```
It is being used on line 205 (`boxed()`). My understanding is that `clippy` would complain if it weren't being used.

gkumbhat replied:
hmm. thats true clippy should complain. But not sure, where am missing, since I don't see reference to `StreamExt` on  205.

https://github.com/foundation-model-stack/fms-guardrails-orchestrator/pull/312/files#diff-1a9212e8f97b31a5e45cc8b28ebcbe726993229a7c05f9f844e266e9b360cb5aR205

Also did a Ctrl+F

mdevino replied:
You won't find `StreamExt` anywhere else in the code, if that's what you were searching for on your Ctrl+F. This is a trait, not a struct.

mdevino commented on tests/streaming_classification_with_gen_nlp.rs:
```diff
@@ -0,0 +1,1212 @@
+            }
+    );
+    assert!(messages[0].input_token_count == mock_tokenization_response.token_count as u32);
+    assert!(messages[0].warnings == Some(vec![DetectionWarning{ id: Some(fms_guardrails_orchestr8::models::DetectionWarningReason::UnsuitableInput), message: Some("Unsuitable input detected. Please check the detected entities on your input and try again with the unsuitable input removed.".into()) }]));
```
I've broken this into two lines, but shouldn't this be configured in the linter/formatter?

gkumbhat reviewed (approved):
left a cleanup comment. But otherwise looks good.

We do need to optimize the test implementation, which can now be done more effectively with the latest release of mocktail and some other strategies we discussed. But that can come up in subsequent PR.

mdevino commented:
As requested internally, I have grouped similar tests. More specifically:

* Updated mocktail to 0.1.3-alpha
* Renamed/grouped tests
* Removed `test_input_detector_whole_doc_no_detections` and `test_input_detector_whole_doc_with_detections` as there are plans to prevent whole_doc_chunker to be used on streaming endpoints.
* Removed `test_input_detector_returns_404`, `test_input_detector_returns_503` and `test_input_detector_returns_non_compliant_message`

cc: @gkumbhat @declark1

declark1 reviewed (approved):
Looks good now @mdevino. Thanks for all of the back-and-forth and your patience while we figure all of this out.

I noticed there are client errors during creation (I think because we don't actually run client health servers), but that isn't a blocker for the tests and can be addressed later.
